File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    bool palidrome(string s, int l, int r){
        while (l < r){
            if (s[l] != s[r]) return false;
            l += 1;
            r -= 1;
        }
        return true;
    }

    void partition(int i, string s, vector<string>& path, vector<vector<string>>& result){
        if (i >= s.length()){
            result.push_back(path);
            return;
        }
        for(int j = i; j < s.length(); j++){
            if (palidrome(s, i, j)){
                path.push_back(s.substr(i, j-i + 1));
                partition(j + 1, s, path, result);
                path.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        vector<string> path;
        partition(0, s, path, result);
        return result;
    }
};

File: 131-palindrome-partitioning/palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        path = []
        result = []

        def palidome(s, l, r):
            while l < r:
                if s[l] != s[r]:
                    return False
                l, r = l + 1, r - 1
            return True

        def backtracking(index):
            if index >= len(s): 
                result.append(path.copy())
                return
            for i in range(index, len(s)):
                if palidome(s, index, i):
                    path.append(s[index: i + 1])
                    backtracking(i + 1)
                    path.pop()

        backtracking(0)
        return result
